File: dcp-solutions/2024-10-29/solution.js
```javascript
class TreeNode {
    constructor(val, left = null, right = null, op = '') {
        this.val = val;
        this.left = left;
        this.right = right;
        this.op = op;
    }
}

function evaluateTree(root) {
    if (!root || root.val === 0) {
        return 0;
    }

    const leftVal = evaluateTree(root.left);
    const rightVal = evaluateTree(root.right);

    switch (root.op) {
        case '+':
            return leftVal + rightVal;
        case '-':
            return leftVal - rightVal;
        case '*':
            return leftVal * rightVal;
        default: // '/'
            if ((leftVal === 0 || rightVal === 0) && root.val !== '1') { // division by zero
                throw new Error('Division by Zero');
            }
            return leftVal / rightVal;
    }
}
```

File: dcp-solutions/2024-10-29/solution.rb
```ruby
class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

def evaluate_tree(root)
  return root.val if root.left.nil? && root.right.nil?

  left_result = evaluate_tree(root.left)
  right_result = evaluate_tree(root.right)

  case root.val
  when '+', '*'
    left_result + right_result
  when '-'
    left_result - right_result
  when '/'
    left_result.to_f / right_result
  end
end
```